framework/src/timesteppers/CompositionDT.C

```diff
@@ -114,38 +199,53 @@ CompositionDT::getSequenceSteppersNextTime()
   Real next_time_to_hit = std::numeric_limits<Real>::max();
   for (auto & tss : time_sequence_steppers)
   {
-    tss->init();
     Real ts_time_to_hit = tss->getNextTimeInSequence();
     if (ts_time_to_hit - _time <= _dt_min)
     {
       tss->increaseCurrentStep();
       ts_time_to_hit = tss->getNextTimeInSequence();
     }
     if (next_time_to_hit > ts_time_to_hit)
+    {
+      _last_time_sequence_stepper = tss;
       next_time_to_hit = ts_time_to_hit;
+    }
   }
   return next_time_to_hit;
 }
 
 Real
 CompositionDT::produceCompositionDT(const std::set<Real> & dts, const std::set<Real> & bound_dts)
 {
-  Real minDT, lower_bound = 0;
+  Real minDT, lower_bound, dt;
+  minDT = lower_bound = dt = 0.0;
   if (!dts.empty())
     minDT = *dts.begin();
   if (!bound_dts.empty())
     lower_bound = *bound_dts.rbegin();
 
+  if (minDT > lower_bound)
+    dt = minDT;
+  else
+  {
+    dt = lower_bound;
+    _last_time_stepper = _bound_last_time_stepper;
+  }
+
   auto ts = getSequenceSteppersNextTime();
 
-  if (ts != 0)
-    return std::min((ts - _time), std::max(minDT, lower_bound));
+  if (ts != 0 && (ts - _time) < dt)
+  {
+    _last_time_stepper = _last_time_sequence_stepper;
+
+    return std::min((ts - _time), dt);
+  }
   else
-    return std::max(minDT, lower_bound);
+    return dt;
 }
 
 std::vector<TimeStepper *>
-CompositionDT::getTimeSteppers()
+CompositionDT::getTimeSteppers() const
```

Comment:
We should filter out this timestepper here, in order to remove that logic everywhere else